test/common/http/conn_manager_impl_test.cc

```diff
@@ -1248,6 +1248,46 @@ TEST_F(HttpConnectionManagerImplTest, PerStreamIdleTimeoutGlobal) {
   EXPECT_EQ(1U, stats_.named_.downstream_rq_idle_timeout_.value());
 }
 
+TEST_F(HttpConnectionManagerImplTest, AccessEncoderRouteBeforeHeadersArriveOnIdleTimeout) {
+  stream_idle_timeout_ = std::chrono::milliseconds(10);
+  setup(false, "");
+
+  EXPECT_CALL(*codec_, dispatch(_)).WillOnce(Invoke([&](Buffer::Instance&) -> void {
+    Event::MockTimer* idle_timer = setUpTimer();
+    EXPECT_CALL(*idle_timer, enableTimer(std::chrono::milliseconds(10)));
+    conn_manager_->newStream(response_encoder_);
+
+    // Expect resetIdleTimer() to be called for the response
+    // encodeHeaders()/encodeData().
+    EXPECT_CALL(*idle_timer, enableTimer(_)).Times(2);
+    EXPECT_CALL(*idle_timer, disableTimer());
+    // Simulate and idle timeout so that the filter chain gets created.
+    idle_timer->callback_();
+  }));
+
+  setupFilterChain(0, 1);
+
+  EXPECT_CALL(*encoder_filters_[0], encodeHeaders(_, _));
+  EXPECT_CALL(*encoder_filters_[0], encodeData(_, _));
+  EXPECT_CALL(*encoder_filters_[0], encodeComplete());
+  EXPECT_CALL(*encoder_filters_[0], onDestroy());
+
+  EXPECT_CALL(response_encoder_, encodeHeaders(_, _));
+  EXPECT_CALL(response_encoder_, encodeData(_, _));
+
+  Buffer::OwnedImpl fake_input;
+  conn_manager_->onData(fake_input, false);
+
+  // This should not be called as we don't have request headers.
+  EXPECT_CALL(*route_config_provider_.route_config_, route(_, _)).Times(0);
+
+  // Under heavy load it is possible that stream timeout will be reached before any headers were
+  // received. Envoy will create a local reply that will go through the encoder filter chain. We
+  // want to make sure that encoder filters get a null route object.
+  auto route = encoder_filters_[0]->callbacks_->route();
```

Comment:
Instead of doing this here, can you potentially make this test a little more realistic by installing a mocker encoder filter and actually calling `route()` from within that filter's callbacks similar to what your filter does? You can see examples below in this file of how to setup a mock filter, it should be pretty simple.